File: 131.分割回文串.cpp
// 回溯，判断是否为回文串是通过字符串左右往中间判断
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> temp;
        dfs(temp, res, s);
        return res;
    }

    void dfs(vector<string> temp, vector<vector<string>>& res, string s){
        if(s.length() == 0){
            res.push_back(temp);
            return;
        }
        for(int i = 0; i < s.length(); ++i){
            if(isPalindrome(s, 0, i)){
                temp.push_back(s.substr(0, i + 1));
                dfs(temp, res, s.substr(i + 1));
                temp.pop_back();
            }
        }
    }

    bool isPalindrome(string s, int start, int end){
        while(start < end) {
            if(s[start] != s[end])
                return false;
            ++start;
            --end;
        }
        return true;
    }
};

// 回溯，判断是否为回文串是通过动态规划标记字符串s那一段属于回文串，即获取是否为回文串的时间复杂度为O(1)
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> temp;
        int len = s.size();
        if(len == 0)
            return res;
        int** isPalindrome = new int* [len];
        for(int i = 0; i < len; ++i)
            isPalindrome[i] = new int[len];
        for(int i = 0; i < len; ++i){
            for(int j = 0; j < len; ++j)
                isPalindrome[i][j] = 0;
        }
        for(int right = 0; right < len; ++right){
            for(int left = 0; left <= right; ++left){
                if(s[left] == s[right] && ((right - left) <= 2 || isPalindrome[left + 1][right - 1]))
                    isPalindrome[left][right] = 1;
            }
        }
        dfs(temp, res, s, 0, isPalindrome);
        return res;
    }

    void dfs(vector<string> temp, vector<vector<string>>& res, string s, int start, int** isPalindrome){
        if(s.size() == start){
            res.push_back(temp);
            return;
        }
        for(int i = start; i < s.length(); ++i){
            if(isPalindrome[start][i]){
                temp.push_back(s.substr(start, i - start + 1));
                dfs(temp, res, s, i + 1, isPalindrome);
                temp.pop_back();
            }
        }
    }
};

File: 动态规划dp/131.分割回文串.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function (s) {
    let n = s.length;
    if (n === 0) {
        return [
            []
        ];
    }
    let res = [];
    // dp[i][j] 表示s[i~j]是否为回文串
    let dp = new Array(n).fill(0).map(() => new Array(n).fill(false));
    for (let i = 0; i < n; ++i) {
        dp[i][i] = true;
    }
    for (let i = n - 2; i >= 0; --i) {
        for (let j = i + 1; j < n; ++j) {
            if (s[i] === s[j]) {
                dp[i][j] = j - i + 1 <= 2 ? true : dp[i + 1][j - 1];
            }
        }
    }

    const dfs = (s, temp, start) => {
        if (s.length === start) {
            res.push([].concat(temp));
            return;
        }
        for (let i = start; i < s.length; ++i) {
            if (dp[start][i]) {
                temp.push(s.substring(start, i + 1));
                dfs(s, temp, i + 1);
                temp.pop();
            }
        }
    }

    dfs(s, [], 0);
    return res;
};